### ISSUE TITLE
Reorg: Data.MonadicStreamFunction.Core

### ISSUE THREAD

**Issue by turion** [ivanperez-keera/dunai#53]
I'd like a reorganisation of Data.MonadicStreamFunction along the following principles:

* `Data.MonadicStreamFunction.Core` implements only core features, to which I'd count:
  * `MSF` definition and its `Arrow` and `Applicative` instances
  * Lifting `MSF` along monad morphisms
  * Reactimation and embedding
  * Reexport of `Control.Arrow` (like Yampa core does)
* Everything else is moved to `Data.MonadicStreamFunction.Util` or other places.
* Modules like `Control.Monad.Trans.MSF.*` and `Data.MonadicStreamFunction.Async` only need to import core. They may use the `MSF` constructor.
* `Data.MonadicStreamFunction.Util` implements utilities using core and possibly other modules. It doesn't need to use the `MSF` constructor.

**Comment by ivanperez-keera** [comment]
Where would the arrow instances go? In `Data.MonadicStreamFunction.Util`?

**Comment by ivanperez-keera** [comment]
We have a number of potentially overlapping tickets:

https://github.com/ivanperez-keera/dunai/issues/53
https://github.com/ivanperez-keera/dunai/issues/48
https://github.com/ivanperez-keera/dunai/issues/26
https://github.com/ivanperez-keera/dunai/issues/16

This issue actually contains two issues: roughly "express things outside the core without the MSF constructor" (https://github.com/ivanperez-keera/dunai/issues/48) and "reorganise module structure".

I would suggest that this one becomes just the module reorganisation, and that we address each issue one at a time (even if they are related). Otherwise, it's really hard to move forward.

**Comment by turion** [comment]
I say `ArrowChoice` and `ArrowLoop` go directly to the `MSF` definition. Agree, this issue is just about reorganisation.

**Comment by ivanperez-keera** [comment]
Do you mean that ArrowChoice should access the constructor, or that it should accompany the definition of MSF?

**Comment by turion** [comment]
Both. (#55)

**Comment by ivanperez-keera** [comment]
There are two implementations of `ArrowChoice` for FRP/Yampa `SF`s. With this, we'd get only one. If we `newtype SF`, then we'd have to re-implement everything in dunai for the newtype to get the other implementation of `ArrowChoice`.

**Comment by turion** [comment]
What are the different `ArrowChoice` implementations? I only see one in our code.

**Comment by ivanperez-keera** [comment]
`ArrowChoice` for SFs needs to "temporarily turn off" a branch and continue down the other.

In implementations based on time deltas, that would mean the clock would not even tick for that branch, meaning that local time would disagree with global time.

This already happens in switch (which bugs me).

In dunai we can offer global time in the monad, so maybe that's the way to address it so that we can provide a default implementation of `ArrowChoice`

**Comment by turion** [comment]
True, important point. But I'd see it as a feature, not a bug. It means that you can "freeze" an `SF` while running the other. There is no way around it. You can't run both `SF`s, since they need input.
You also can't delay the last sample since then one branch won't have input on the first sample. If you  supply initialisation values, you could run both `SF`s in parallel, but that can be implemented using only `feedback` and `Arrow`. And it's not exactly `ArrowChoice`.

So I still only see one valid `ArrowChoice` instance. Is there another one? If not, let's just put it in the same module as the `MSF` definition.

**Comment by ivanperez-keera** [comment]
In Yampa, at least, you can. Imagine that you get two samples, one at time `t_i1`, and another at time `t_i2`. Let's assume `t_i0 = 0`

Imagine also that they are processed by two different SFs by means of `ArrowChoice`. The first one would receive the sample with delta `t_i1`, and the second one would receive the sample with time `t_i2 - t_i1`. You could also accumulate the time, add it, and then have the first sample received with delta `t_i1`, and the second with `t_i2`. The question I have to think about is: what would that mean? Would that make something inconsistent, or less inconsistent?

**Comment by turion** [comment]
Ah, I understand what you mean. That's a very particular thing to `Num a => Reader a`. I think you can recover the other `ArrowChoice` instance like this:

* Imagine we had a function `withReaderS :: MSF m r1 r2 -> MSF (ReaderT r1 m) a b -> MSF (ReaderT r2 m) a b`. (It's no problem to implement it with dunai.)
* Implement `differenceS :: Num a => MSF m a a` as `differenceS = id - delay 0`, using the `Num` instance for `MSF`s. If I'm right we should have that `sumS >>> differenceS == id`.
* With the synonym `type SF a b = MSF (Reader Double) a b`, use this to convert between absolute time and differential time: `withDiffTime :: SF a b -> SF a b`, `withDiffTime = withReaderS differenceS`, and analogously `withAbsoluteTime :: withReaderS sumS`.
* Now your new `SF`-specific `ArrowChoice` instance would be `leftSF :: SF a b -> SF (Either a c) (Either b c)` given by `leftSF = withDiffTime . left . withAbsoluteTime`.

**Comment by turion** [comment]
I'd look at it like that: Yampa has no `ArrowChoice` instance, so there is no need for one in bearriver. One could implement one as you describe it, but that's very specific to `SF`s and a special case of the general `ArrowChoice` (as shown above).

I think it's not worth an orphan instance. That would mean that we'd have two different `ArrowChoice` instances flying around, and depending on which module got imported somewhere, a different instance is selected. That would be very confusing. Imagine the following:

1. We keep the `ArrowChoice` instance as an orphan and add the other `SF`-specific instance to bearriver.
2. Someone uses dunai and gets used to the way `ArrowChoice` works there.
3. That user discovers bearriver, recognizes that `SF` is just a type synonym, notices the `ArrowChoice` instance and thinks: "Ah, good, so here the `ArrowChoice` instance is already imported."
4. Really strange bugs happen, time goes faster or slower than the user thinks, until they bother reading the code _which_ instance we actually imported. Not good.

The other option is having the `newtype`. It's a nuisance, but not such a big one. A lot of names are different in Yampa than they are in dunai anyways (`delay` vs. `iPre`, `feedback` vs. `loopPre` etc.), and we should keep the dunai names because they are better. And for the names that stay the same, it's just a matter of writing two trivial lines per name. (And all instances we want to keep can be copied with `GeneralisedNewtypeDeriving`.

The third option is of course using [rhine](https://github.com/turion/rhine), since it solves your problem as well :)

**Comment by turion** [comment]
To collect the current status of this issue:

# Uncontroversial proposed changes

* `Data.MonadicStreamFunction.Core` implements only core features, to which I'd count:
  * `MSF` definition and its `Arrow` and `Applicative` instances
  * Lifting `MSF` along monad morphisms
  * Reactimation and embedding
  * Reexport of `Control.Arrow` (like Yampa core does)
* Everything else is moved to `Data.MonadicStreamFunction.Util` or other places.

# Controversial proposed changes

* Move `ArrowChoice` into `Core`.

# Expected benefit

* `Core` is small.
* Modules like `Control.Monad.Trans.MSF.*` and `Data.MonadicStreamFunction.Async` only need to import core.
* `Data.MonadicStreamFunction.Util` implements utilities using core and possibly other modules.

# Actions to be taken now

* Decide about `ArrowChoice`. I say it should go to `Core`. If you disagree, let's move this question into a separate issue so I can resolve the rest.

**Comment by turion** [comment]
Possibly, `ArrowPlus` and `ArrowLoop` are also controversial, although I can't see other definitions. See #55.

**Comment by turion** [comment]
@ivanperez-keera can I assume that you need more time/research to decide on the `ArrowChoice` matter, and we can resolve the rest of the module reorganisation?

**Comment by ivanperez-keera** [comment]
> Yampa has no ArrowChoice instance, so there is no need for one in bearriver.

It does. It was added in a branch and will be added to the main tree soon.

**Comment by ivanperez-keera** [comment]
I'm actually also partially unhappy about the instance of `ArrowLoop`.

I could be wrong, but I think it could potentially be used to define non-causal stream functions. I'd rather keep it out until we can check that that's not the case.

**Comment by ivanperez-keera** [comment]
What would be the "Lifting MSF along monad morphisms" facility that would need to be included in `Core`.

Ivan

**Comment by ivanperez-keera** [comment]
> I mean specifically: which specific types and functions would be included in Core due to including 'Lifting MSFs along monad morphisms'.

The answer may be obvious to you, but there's https://github.com/ivanperez-keera/dunai/issues/26.

**Comment by turion** [comment]
Yes, `ArrowLoop` can in principle hang up. If you loop e.g. `id` in the `Identity` monad, that should already hang. I think we talked about this at some point. People should use `feedback` wherever possible. (The deeper issue is that Haskell doesn't have a termination checker, unlike Agda.)

**Comment by ivanperez-keera** [comment]
> Yes, ArrowLoop can in principle hang up.

This is not exactly what I mean. I mean that it's possible to make present outputs depend on future inputs.

**Comment by ivanperez-keera** [comment]
But what you say is also true.

**Comment by turion** [comment]
I think it's by definition of `MSF` not possible that present outputs depend on future inputs.

**Comment by ivanperez-keera** [comment]
> I think it's by definition of MSF not possible that present outputs depend on future inputs.

I think this is generally not the case with `MonadFix` and if we give people access to the `MSF` constructor.

**Comment by turion** [comment]
> What would be the "Lifting MSF along monad morphisms" facility that would need to be included in Core
Basically, all the lifting things that already are in Core.

The long term goal would be to add the relevant parts of `Control.Monad.Trans.MSF.GenLift`, but that needs a little bit research still.

**Comment by ivanperez-keera** [comment]
There's also https://github.com/ivanperez-keera/dunai/pull/84 and #16 , which in some way address part of those changes by adding a function `hoistGen` to `Core` and defining some exception handling MSF combinators in terms of it.

**Comment by turion** [comment]
> Ok. So, if we do not add ArrowPlus, ArrowChoice or ArrowLoop, specifically, how would the module Core change?

Switching and a few other primitives are moved to `Util`, which resolves an import cycle which allows #69 to move forward.

**Comment by turion** [comment]
> I think this is generally not the case with MonadFix and if we give people access to the MSF constructor.
Maybe I don't understand `MonadFix` properly. Anyways, we both see problems with `ArrowLoop`. Could we still add `ArrowLoop`, but say in the documentation of the instance, that it's dangerous? (No worries if not)

Can we still add `ArrowPlus`? It's not causing problems, right?

> >  Yampa has no ArrowChoice instance, so there is no need for one in bearriver.
> It does. It was added in a branch and will be added to the main tree soon.
Oh ok. And does time go on running for both `sf`s there?

**Comment by ivanperez-keera** [comment]
I have no problem with having 'ArrowZero' and 'ArrowPlus' in either place. Can they be implemented without using the constructors?

**Comment by turion** [comment]
I think they can't. Consider the two type checking implementations of `zeroArrow`: `arrM $ const mzero` and `MSF $ const mzero`. They probably are the same, although it's not completely obvious to me. But I see no way of implementing `<+>` without the constructor. Note that `&&&` already sequentialises the side effects of `sf1` and `sf2`, but we really want to parallelise them inside the monad.

There is a generalisation like `(forall a . m1 a -> m2 a -> m3 a) -> MSF m1 b c -> MSF m2 b c -> MSF m3 b c`, which is something like an `Applicative` in the category of monads, but I don't know whether we'll ever need that (since it doesn't specialise to anything else we use).

**Comment by turion** [comment]
Here's another resolution to the problem of `MSF` and bearriver's `SF` having different `ArrowChoice` instances, following Oleg Kiselyov (http://okmij.org/ftp/tagless-final/). Just abstract `Data.MonadicStreamFunction.Core` as a type class. This really just amounts to:

```haskell
class ArrowM arr where
  arrM :: (a -> m b) -> arr m a b

class ArrowLoopPre arr where
  feedback :: c -> arr (a, c) (b, c) -> arr a b
```
* Then the rest of core, and indeed  can be written over this type class, i.e. `iPre :: ArrowLoopPre arr => a -> arr a a`.
* Then write instances `instance ArrowM MSF` and `instance Monad m => ArrowLoopPre (MSF m)` using the constructor. The only other thing where we need the constructor is reactimation. (And lifting/exception handling, which deserve their own type classes.)
* When declaring bearriver's `SF`s as newtypes, just reuse all `ArrowM`, `ArrowPlus` etc. instances via newtype deriving. Write a new `ArrowChoice`. All of dunai automatically works for `SF`s as well, because it's written for type classes and not a specific type. And everything we'll add will also immediately work.

This would also be advantageous for rhine (https://github.com/turion/rhine/issues/20).

**Comment by ivanperez-keera** [comment]
This is what I've been planning to do next.
